### PR TITLE
[TASK] Media browser performance improvements

### PR DESCRIPTION
**Pull request by aertmann**
Optimizes rendering performance of the media browser/module by
re-using modals, optimizing and using count queries.

Brings 0,5x-3x+ speed improvement depending on the selected view
and amount of assets and tags/asset collections visible.

No functional difference.

Additionally fixes a couple of bugs in the asset repository.

Resolves: NEOS-1698

### CODE REVIEW

**Comment by aertmann** [comment]
:+1:

**Comment by aertmann** [comment]
Example of optimized query

````
#SELECT count(DISTINCT t0_.persistence_object_identifier) AS sclr0 FROM typo3_media_domain_model_asset t0_ LEFT JOIN typo3_media_domain_model_audio t1_ ON t0_.persistence_object_identifier = t1_.persistence_object_identifier LEFT JOIN typo3_media_domain_model_document t2_ ON t0_.persistence_object_identifier = t2_.persistence_object_identifier LEFT JOIN typo3_media_domain_model_image t3_ ON t0_.persistence_object_identifier = t3_.persistence_object_identifier LEFT JOIN typo3_media_domain_model_imagevariant t4_ ON t0_.persistence_object_identifier = t4_.persistence_object_identifier LEFT JOIN typo3_media_domain_model_video t5_ ON t0_.persistence_object_identifier = t5_.persistence_object_identifier WHERE (EXISTS (SELECT 1 FROM typo3_media_domain_model_assetcollection_assets_join t6_ INNER JOIN typo3_media_domain_model_assetcollection t7_ ON t6_.media_assetcollection = t7_.persistence_object_identifier WHERE t6_.media_asset = t0_.persistence_object_identifier AND t7_.persistence_object_identifier = '01bdb83b-fd80-cd8b-97e7-ff458d82be5e'))
````
vs.
````
SELECT count(a.persistence_object_identifier) c FROM typo3_media_domain_model_asset a LEFT JOIN typo3_media_domain_model_assetcollection_assets_join collectionmm ON a.persistence_object_identifier = collectionmm.media_asset WHERE collectionmm.media_assetcollection = '01bdb83b-fd80-cd8b-97e7-ff458d82be5e'
````

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php [hunk 1]
```diff
@@ -94,9 +95,9 @@ public function countByTag(\TYPO3\Media\Domain\Model\Tag $tag, AssetCollection $
...
         } else {
-            $queryString = 'SELECT count(a.persistence_object_identifier) c FROM typo3_media_domain_model_asset a LEFT JOIN typo3_media_domain_model_asset_tags_join tagmm ON a.persistence_object_identifier = tagmm.media_asset LEFT JOIN typo3_media_domain_model_assetcollection_assets_join collectionmm ON a.persistence_object_identifier = collectionmm.media_asset WHERE tagmm.media_tag = ? AND collectionmm.media_assetcollection = ?';
+            $queryString = 'SELECT count(a.persistence_object_identifier) c FROM typo3_media_domain_model_asset a LEFT JOIN typo3_media_domain_model_asset_tags_join tagmm ON a.persistence_object_identifier = tagmm.media_asset LEFT JOIN typo3_media_domain_model_assetcollection_assets_join collectionmm ON a.persistence_object_identifier = collectionmm.media_asset WHERE tagmm.media_tag = ? AND collectionmm.media_assetcollection = ? AND a.dtype != "typo3_media_imagevariant"';
         }
 
         $query = $this->entityManager->createNativeQuery($queryString, $rsm);
```
Comment (kitsunet) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php]: Does this query work with all supported DBs?

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php [hunk 2]
```diff
@@ -123,9 +124,13 @@ public function findAll(AssetCollection $assetCollection = null)
      */
     public function countAll()
     {
-        $query = $this->createQuery();
-        $this->addImageVariantFilterClause($query);
-        return $query->count();
+        $rsm = new \Doctrine\ORM\Query\ResultSetMapping();
+        $rsm->addScalarResult('c', 'c');
+
+        $queryString = 'SELECT count(persistence_object_identifier) c FROM typo3_media_domain_model_asset WHERE dtype != "typo3_media_imagevariant"';
```
Comment (kitsunet) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php]: same question here

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php [hunk 3]
```diff
@@ -151,11 +156,20 @@ public function findUntagged(AssetCollection $assetCollection = null)
...
+        $rsm = new \Doctrine\ORM\Query\ResultSetMapping();
+        $rsm->addScalarResult('c', 'c');
+
+        if ($assetCollection === null) {
+            $queryString = 'SELECT count(a.persistence_object_identifier) c FROM typo3_media_domain_model_asset a LEFT JOIN typo3_media_domain_model_asset_tags_join tagmm ON a.persistence_object_identifier = tagmm.media_asset WHERE tagmm.media_asset IS NULL AND a.dtype != "typo3_media_imagevariant"';
+        } else {
+            $queryString = 'SELECT count(a.persistence_object_identifier) c FROM typo3_media_domain_model_asset a LEFT JOIN typo3_media_domain_model_asset_tags_join tagmm ON a.persistence_object_identifier = tagmm.media_asset LEFT JOIN typo3_media_domain_model_assetcollection_assets_join collectionmm ON a.persistence_object_identifier = collectionmm.media_asset WHERE tagmm.media_asset IS NULL AND collectionmm.media_assetcollection = ? AND a.dtype != "typo3_media_imagevariant"';
+        }
+
+        $query = $this->entityManager->createNativeQuery($queryString, $rsm);
```
Comment (kitsunet) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php]: and here

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php [hunk 4]
```diff
@@ -151,11 +156,20 @@ public function findUntagged(AssetCollection $assetCollection = null)
      */
     public function countUntagged(AssetCollection $assetCollection = null)
     {
-        $query = $this->createQuery();
-        $query->matching($query->isEmpty('tags'));
-        $this->addImageVariantFilterClause($query);
```
Comment (kitsunet) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php]: Is this still in use then?

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php [hunk 5]
```diff
@@ -151,11 +156,20 @@ public function findUntagged(AssetCollection $assetCollection = null)
...
-        $this->addAssetCollectionToQueryConstraints($query, $assetCollection);
```
Comment (kitsunet) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php]: Is this still in use?

**Comment by kitsunet** [comment]
left some comments, the rest is fine I guess.

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php (hunk 4 above)
Comment (aertmann) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php]: countUntagged? it's used in the controller..

Comment (kitsunet) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php]: no, the methods `addImageVariantFilterClause` and `addAssetCollectionToQueryConstraints`

**Comment by aertmann** [comment]
@kitsunet good catch, checked the query on postgres and it fails.. will push follow up

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php (hunk 4 above)
Comment (aertmann) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php]: yes sure, just not in the count methods

Comment (aertmann) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php]: https://github.com/aertmann/neos-development-collection/blob/media-browser-performance-improvements/TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php#L80

**Comment by aertmann** [comment]
:+1:

**Comment by aertmann** [comment]
btw. since there's no functional changes, simply verify that the counts still work and the modal windows (delete asset, delete tag, delete collection) appear and work..

**Comment by aertmann** [comment]
for reference, this changed the response time from 35 s to 5 s with 6000 assets and ~350 tags

**Comment by kitsunet** [comment]
Good stuff, thanks a lot :+1:

**Comment by aertmann** [comment]
thanks

**Comment by gerhard-boden** [comment]
i saw this has already been merged, but i still tried to "manually merge" this changes into my 2.0.4 test system (since i got around a thousand assets there to test the behaviour and performance) but unluckily for me these changes seem to depend on other changes not merged into my 2.0.4 instance yet, since i get:
`#1337700024: f:format.identifier expects an object, array given.`

**Pull request merged**